### ISSUE TITLE
Allow suggestions/feedback in comments

### ISSUE THREAD

**Issue by eternoendless** [Reviewable/Reviewable#671]
Github implemented a while ago a very useful feature called "suggestions" or "feedback" ([see here](https://help.github.com/en/articles/incorporating-feedback-in-your-pull-request)), where a reviewer can add a suggestion block that can be easily accepted and merged by the PR author afterwards.

It would be nice if this feature was available on Reviewable as well :)

**Comment by pkaminski** [comment]
Hmm, I thought there was already an issue for this but can't find it, so I guess this will be it.

I agree this would be very nice to have, but the implementation turns out to be somewhat tricky both because it doesn't fit cleanly with Reviewable's UI, and because GitHub's API doesn't make merging suggestions easy.

**Comment by maxfischer2781** [comment]
Is there some way to have Reviewable create the appropriate suggestion comment on GitHub so that we can merge via GitHub? We've tried annotating a code block with ``suggestion`` (as GitHub does) in Reviewable but that did not translate to a proper review comment.

**Comment by pkaminski** [comment]
Yeah, that won't work -- sorry.  :(  A suggestion needs to be in its own (GitHub) comment, but Reviewable needs to batch-post comments due to fundamental flaws in GitHub's API as explained in #437.  In fact, I'm going to mark this issue as blocked since it doesn't make much sense to implement suggestion merging if we can't support suggestion posting.

**Comment by Cactucs** [comment]
I would disagree that it it is blocked by GitHub's API. For me it would be useful also without ability to commit the changes, but when I want to suggest a small change (such as a fix for a typo) it would be much easier with this feature.

When using GitHub's review a I spot a typo I click the suggest button and it creates a code block where I can easily fix it. Then it is shown to the PR author in an understandable way. 

For me, this is the feature I miss when I use Reviewable (which is otherwise much better).

**Comment by pkaminski** [comment]
Merging suggestions in Reviewable is feasible, it's just a lot of work to implement it.  Creating suggestions that interoperate with GitHub is currently precluded by #437 and building our own parallel and incompatible system is not going to fly.

Incidentally, I'm really sad that after promising to build their product on the public API, GitHub is still adding features implemented in a proprietary layer.  (Calling @nat...)

**Comment by pkaminski** [comment]
No real progress on this, but I found the duplicate issue:  #415.

**Comment by pkaminski** [comment]
I've cracked and started work on this feature, with two caveats:
1. Suggestions will not interoperate with GitHub due to the lack of any API.
2. You won't be able to commit a suggestion directly from Reviewable.  I think that committing snippets of code written in a text field and never actually executed (or even linted or type-checked!) is a pretty bad idea anyway, and also doesn't play well with an amend / force push workflow.  Instead, I intend to offer a "copy and jump to line in editor" button.

If somebody wants to argue for direct commits I'm open to hearing your arguments!

**Comment by eternoendless** [comment]
We use suggestion commits extensively on GitHub, they are checked by CI after they are committed. A big plus is that suggested changes are committed as co-authored by the person who made it and the person who accepts them, so attribution is better managed than by copy-pasting.

**Comment by schreter** [comment]
Independent of the possibility to commit the suggestion, it would be already quite useful to *give* the suggestion by showing a small diff like in github. I typically used it to make suggestions in documentation, not the code. Now I use a workaround by selecting text and then modifying the quoted text.

Maybe it would be easiest to simply interpret a modified quote as a suggestion and display it as a diff instead of a quote.

How to later *apply* those suggestions is another story. But at least we'd have the possibility to easily add suggestions.

**Comment by pkaminski** [comment]
Yep, I'm actually pretty close to having this working.  But the suggestion code editor is so nice that I'm considering being ambitious and leveraging it for all codeblocks...  In any case, this will be the next major feature to be released.

**Comment by schreter** [comment]
> Yep, I'm actually pretty close to having this working.

Thanks, looking forward to it :-). Any ETA yet?

**Comment by pkaminski** [comment]
Hey folks, I think we're just about ready to open this up for beta testing.  If you'd like to help us out, please post a list of usernames here (or send to support@reviewable.io) and I'll add them to the whitelist.  Whitelisted users will be able to create and publish suggestions, but everyone will be able to see them once published.

There are two known limitations at the moment:
1. There's no keyboard shortcut assigned to creating a suggestion because I can't put it behind the whitelist.  I intend to assign `cmd+shift+e` by default.
2. The jump to line function (to see where a suggestion came from, as it doesn't have to be near the comment) is not yet implemented.  This ties into #147 and will need some infrastructure built up from that side first, so it will take a while yet.  I'll probably hide it to avoid confusing people before releasing this but I wanted to leave the UI in for now so you'll be able to see how things are intended to work.

Once you're in, I'd love some feedback on how easy it was to discover and use the feature, and of course reports of any bugs you run into.  Thanks in advance!

**Comment by ggould-tri** [comment]
I'd love to join the test!  @ggould-tri is the username.

**Comment by pkaminski** [comment]
Cool, you're in @ggould-tri.

I forgot to mention one more limitation:  we're working on a more compact diff style for minor suggestions, so we don't need the full red line / green line of a standard unidiff for each change.

**Comment by pkaminski** [comment]
Nobody else wants to give this a try?  @eternoendless, @VojtaStanek, @maxfischer2781, @chandlerc, @melink14, @alexspeller, @mendz

**Comment by alexjurkiewicz** [comment]
I would like to try, I'm @alexjurkiewicz 👋

**Comment by pkaminski** [comment]
Cool, you're in @alexjurkiewicz!

**Comment by ls-jason-rossiter** [comment]
I would like to give this a try too! @ls-jason-rossiter

**Comment by pkaminski** [comment]
Sweet, you're in too!

**Comment by ls-rajan-rauniyar** [comment]
I would love to try this too. Thanks  @ls-rajan-rauniyar

**Comment by pkaminski** [comment]
Sure thing, done.

**Comment by ls-omid-vahdani** [comment]
I would like to give this a try too. @ls-omid-vahdani

**Comment by pkaminski** [comment]
And done!

**Comment by ggould-tri** [comment]
One broken-glasses this morning:
![image](https://user-images.githubusercontent.com/17596413/144865965-ec689975-ef9f-4c1d-b590-ada8fa6f4626.png)
(maybe the button should be grayed out for unknown languages?)
(let me know if you want bug reports in new issues rather than here)

**Comment by pkaminski** [comment]
Thanks -- fixed, everything should work again.  This was happening because our highlightjs configuration was out of sync between the worker and the main client.  Apologies for the broken review in the meantime!

Reporting errors in this issue is fine.  👍

**Comment by facugaich** [comment]
@pkaminski Could you add me to the whitelist? Thanks @facugaich

**Comment by pkaminski** [comment]
As it happens, we've decide that this feature is stable and useful enough as-is to release immediately, while we work to further improve the diffing.  It's available to everyone right now!
